File: src/ImputedAndKnownFamilyRef.cpp
#include <algorithm>

#include "../include/ImputedAndKnownFamilyRef.h"
#include "../include/ImputedAndKnownFamily.h"
#include "../include/VCFImputable.h"
#include "../include/Pedigree.h"
#include "../include/KnownFamily.h"
#include "../include/OptionSmall.h"
#include "../include/RefCommon.h"
#include "../include/common.h"

using namespace std;


//////////////////// ImputedAndKnownFamilyRef ////////////////////

VCFGeno *ImputedAndKnownFamilyRef::impute(
									const VCFSmall *orig_vcf,
									const VCFGeno *phased_vcf,
									const vector<vector<int>>& ref_haps,
									const vector<const KnownFamily *>& families,
									const STRVEC& non_imputed_parents,
									const OptionSmall& op) {
	const size_t	N = families.size();
	if(N == 0)
		return NULL;
	
	vector<VCFImputable *>	vcfs(N);
	for(size_t i = 0; i < N; ++i) {
		const KnownFamily	*family = families[i];
		const auto&	samples = family->get_samples();
		auto	*vcf = VCFGeno::extract_samples(samples, orig_vcf);
		auto	records = RefCommon::merge_family_records(phased_vcf,
															vcf, samples);
		const bool	is_mat_imputed = std::find(non_imputed_parents.begin(),
											   non_imputed_parents.end(),
											   family->get_pat())
										!= non_imputed_parents.end();
		vcfs[i] = ImputedAndKnownFamily::create_family_vcf(
												family, records, is_mat_imputed,
												N, ref_haps, orig_vcf, op);
		delete vcf;
	}
	
	// Small VCFs are heavy to process, so it will be parallelized.
	VCFImputable::impute_VCFs(vcfs, op.num_threads);
	cout << N << " families whose one parent is imputed and the other parent"
									<< " is known have been imputed." << endl;
	vector<const VCFGenoBase *>	vcfs1(vcfs.begin(), vcfs.end());
	auto	*new_vcf = VCFGeno::join(vcfs1, orig_vcf->get_samples());
	Common::delete_all(vcfs);
	return new_vcf;
}

File: python/ImputedAndKnownFamilyRef.py
from __future__ import annotations

# coding: utf-8
# ImputedAndKnownFamilyRef.py
# リファレンスがあって、
# 片親がphasingされて片親は分っているがphasingされていない家系を補完する

from typing import Optional, Sequence

from VCF import VCFSmall
from VCFGeno import VCFGenoBase, VCFGeno
from VCFFamily import *
import ImputedAndKnownFamily
import RefCommon
from OptionSmall import OptionSmall


#################### ImputedAndKnownFamilyRef ####################

def impute(orig_vcf: VCFSmall, phased_vcf: VCFGeno,
									ref_haps: list[list[int]],
									families: list[Family],
									non_imputed_parents: list[str],
									op: OptionSmall) -> Optional[VCFGenoBase]:
	vcfs: list[VCFGenoBase] = []
	for family in families:
		vcf = VCFGeno.extract_samples(family.samples(), orig_vcf);
		records = RefCommon.merge_family_records(phased_vcf,
													vcf, family.samples())
		is_mat_imputed = family.pat in non_imputed_parents
		vcf_family = ImputedAndKnownFamily.create_family_vcf(
										family, records, is_mat_imputed,
										len(families), ref_haps, orig_vcf, op)
		vcf_family.impute()
		vcfs.append(vcf_family)
	
	if not vcfs:
		return None
	
	print("%d families whose one parent is imputed and the other parent is"
									" known have been imputed." % len(vcfs))
	new_vcf = VCFGeno.join(vcfs, orig_vcf.samples)
	return new_vcf

__all__ = ['impute']
